cmd/controller-manager/controller-manager.go

```diff
@@ -0,0 +1,85 @@
+package main
+
+import (
+	"context"
+	"flag"
+	"os"
+	"os/signal"
+	"syscall"
+
+	"github.com/spf13/cobra"
+	"github.com/spf13/viper"
+
+	bucketcontroller "github.com/kubernetes-sigs/container-object-storage-interface-api/controller"
+	"github.com/kubernetes-sigs/container-object-storage-interface-controller/pkg/bucketrequest"
+
+	"github.com/golang/glog"
+)
+
+var cmd = &cobra.Command{
+	Use:           "controller-manager",
+	Short:         "central controller for managing bucket* and bucketAccess* API objects",
+	SilenceErrors: true,
+	SilenceUsage:  true,
+	RunE: func(c *cobra.Command, args []string) error {
+		return run(c.Context(), args)
+	},
+	DisableFlagsInUseLine: true,
+}
+
+var kubeConfig string
+
+func init() {
+	viper.AutomaticEnv()
+
+	cmd.PersistentFlags().AddGoFlagSet(flag.CommandLine)
+	flag.Set("logtostderr", "true")
+
+	strFlag := func(c *cobra.Command, ptr *string, name string, short string, dfault string, desc string) {
+		c.PersistentFlags().
+			StringVarP(ptr, name, short, dfault, desc)
+	}
+	strFlag(cmd, &kubeConfig, "kube-config", "", kubeConfig, "path to kubeconfig file")
+
+	hideFlag := func(name string) {
+		cmd.PersistentFlags().MarkHidden(name)
+	}
+	hideFlag("alsologtostderr")
+	hideFlag("log_backtrace_at")
+	hideFlag("log_dir")
+	hideFlag("logtostderr")
+	hideFlag("master")
+	hideFlag("stderrthreshold")
+	hideFlag("vmodule")
+
+	// suppress the incorrect prefix in glog output
+	flag.CommandLine.Parse([]string{})
+	viper.BindPFlags(cmd.PersistentFlags())
+
+}
+
+func main() {
+	if err := cmd.Execute(); err != nil {
+		glog.Fatal(err.Error())
+	}
+
+	var cancel context.CancelFunc
+
+	_, cancel = context.WithCancel(cmd.Context())
+	sigs := make(chan os.Signal, 1)
+	signal.Notify(sigs, syscall.SIGINT, syscall.SIGTERM)
+
+	go func() {
+		<-sigs
+		cancel()
+	}()
```

Comment:
All of this comes after we ran `cmd.Execute`, so it seems unused whatsoever?

Reply:
if you are saying that this code should go before cmd.Execute then we will have to create the context like we did before.

Reply:
We need to create it, but it shouldn't be a global, can remain within `main` :smiley: Something like the following (from memory):

```go
func main() {
    ctx, cancel := context.WithCancel(context.Background())
    defer cancel()  // Just in case

    sigs := make(chan os.Signal, 1)
    signal.Notify(sigs, syscall.SIGINT, syscall.SIGTERM)

    go func() {
        <-sigs
        cancel()
    }()

    if err := cmd.ExecuteContext(ctx); err != nil {
        glog.Fatal(err.Error())
    }
}
```